include/StateChecker.php

```diff
@@ -0,0 +1,322 @@
+     *
+     * @var array
+     */
+    protected $hashes;
+    
+    /**
+     *
+     * @var array
+     */
+    protected $traces;
+    
+    /**
+     *
+     * @var integer
+     */
+    protected $memoryLimit;
+    
+    /**
+     * 
+     * @throws StateCheckerException
+     */
+    public function __construct()
+    {
+        if (!$this->db = DBManagerFactory::getInstance()) {
+            throw new StateCheckerException('DBManagerFactory get instace failure');
+        }
+        if (!($this->db instanceof MysqliManager)) {
+            throw new StateCheckerException('Incompatible DB type, only supported: mysqli');
+        }
+        $this->resetHashes();
+        $this->resetTraces();
+        
+        if (StateCheckerConfig::get('redefineMemoryLimit')) {
+            $this->memoryLimit = ini_get('memory_limit');
+            ini_set('memory_limit', -1);
+        }
+                
+        if (StateCheckerConfig::get('autoRun')) {
+            $this->getStateHash();
+        }
+    }
+    
+    /**
+     * 
+     * @return array traces
+     * @throws StateCheckerException
+     */
+    public function getTraces()
+    {
+        if (StateCheckerConfig::get('saveTraces')) {
+            throw new StateCheckerException('Trace information is not saved, use StateCheckerConfig::get(\'saveTraces\') as true');
+        }
+        return $this->traces;
+    }
+    
+    /**
+     * 
+     */
+    public function __destruct()
+    {
+        if (StateCheckerConfig::get('redefineMemoryLimit')) {
+            ini_set('memory_limit', $this->memoryLimit);
+        }
+    }
+    
+    /**
+     * resetTraces
+     */
+    protected function resetTraces()
+    {
+        $this->traces = [];
+    }
+
+    /**
+     * resetHashes
+     */
+    protected function resetHashes()
+    {
+        $this->hashes = [];
+    }
+    
+    /**
+     * 
+     * @param string $key
+     * @return boolean
+     */
+    protected function isDetailedKey($key)
+    {
+        $detailedKey = preg_match('/\w+\:\:/', $key);
+        return $detailedKey;
+    }
+    
+    /**
+     * 
+     * @param string $hash
+     * @param string $key
+     * @return boolean
+     */
+    protected function checkHash($hash, $key)
+    {
+        $detailedKey = $this->isDetailedKey($key);
+        $needToStore = !$detailedKey || ($detailedKey && StateCheckerConfig::get('storeDetails'));
+        
+        if (!isset($this->hashes[$key])) {
+            if ($needToStore) {
+                $this->hashes[$key] = $hash;
+            }
+        }
+        if ($needToStore) {
+            $match = $this->hashes[$key] == $hash;
+        } else {
+            $match = true;
+        }
+        return $match;
+    }
+    
+    /**
+     * 
+     * @param mixed $data should be serializable
+     * @param string $key
+     * @return string
+     * @throws StateCheckerException
+     */
+    protected function getHash($data, $key)
+    {
+        if (!$serialized = serialize($data)) {
```

Comment:
Assigning or comparison? Please extract it to avoid confusion.

Reply:
Assign. one `=` means assign.